receiver/shopifyreceiver/trace_receiver.go

```diff
@@ -0,0 +1,282 @@
+	var err = errAlreadyStopped
+	sr.stopOnce.Do(func() {
+		err = sr.server.Close()
+	})
+	return err
+}
+
+func (sr *ShopifyReceiver) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	switch {
+	case r.URL.Path == "/":
+		sr.ingest(w, r)
+	default:
+		http.NotFound(w, r)
+	}
+}
+
+func extractKind(span *tracing.Span) tracepb.Span_SpanKind {
+	switch span.GetTags()["span.kind"] {
+	case "client":
+		return tracepb.Span_CLIENT
+	case "server":
+		return tracepb.Span_SERVER
+	default:
+		return tracepb.Span_SPAN_KIND_UNSPECIFIED
+	}
+}
+
+func tagsToTraceAttributes(tags map[string]string) *tracepb.Span_Attributes {
+	if len(tags) == 0 {
+		return nil
+	}
+	attributes := make(map[string]*tracepb.AttributeValue, len(tags))
+	for k, v := range tags {
+		if k == "span.kind" || k == "error" {
+			continue
+		}
+		attributes[k] = &tracepb.AttributeValue{
+			Value: &tracepb.AttributeValue_StringValue{
+				StringValue: &tracepb.TruncatableString{
+					Value: v}}}
+	}
+	if len(attributes) == 0 {
+		return nil
+	}
+	return &tracepb.Span_Attributes{AttributeMap: attributes}
+}
+
+func extractLinks(span *tracing.Span) *tracepb.Span_Links {
+	links := span.GetLinks().GetLink()
+	if len(links) == 0 {
+		return nil
+	}
+	ls := make([]*tracepb.Span_Link, 0, len(links))
+	for _, l := range links {
+		if l == nil {
+			continue
+		}
+		ls = append(ls, &tracepb.Span_Link{
+			TraceId: l.TraceId,
+			SpanId:  l.SpanId,
+			Type:    tracepb.Span_Link_Type(l.Type),
+		})
+	}
+	return &tracepb.Span_Links{Link: ls}
+}
+
+func extractStatus(span *tracing.Span) *tracepb.Status {
+	// TODO: the 'error' tag from OpenTracing is too simplistic. We could map 'http.status_code' here instead.
+	if span.GetTags()["error"] != "true" {
+		return nil
+	}
+	return &tracepb.Status{
+		Code:    2,
+		Message: "UNKNOWN",
+	}
+}
+
+func toTraceSpans(spans tracing.Spans) (ocSpans []*tracepb.Span, invalidSpans int, err error) {
+	ocSpans = make([]*tracepb.Span, 0, len(spans.Spans))
+	for _, span := range spans.Spans {
+		if span == nil {
+			invalidSpans++
+			continue
+		}
+		traceID, err := hex.DecodeString(span.TraceId)
+		if err != nil {
+			invalidSpans++
+			continue
+		}
+		parentSpanID := span.GetChildOf()
+		if parentSpanID == 0 {
+			parentSpanID = span.GetFollowsFrom()
+		}
+		ocSpans = append(ocSpans, &tracepb.Span{
+			TraceId:      traceID[:16],
+			SpanId:       tracetranslator.UInt64ToByteSpanID(span.Id),
+			ParentSpanId: tracetranslator.UInt64ToByteSpanID(parentSpanID),
+			Name:         &tracepb.TruncatableString{Value: span.Operation},
+			Kind:         extractKind(span),
+			StartTime:    span.Start,
+			EndTime:      span.End,
+			Attributes:   tagsToTraceAttributes(span.Tags),
+			Links:        extractLinks(span),
+			Status:       extractStatus(span),
+		})
+	}
+	return
+}
+
+func (sr *ShopifyReceiver) ingest(w http.ResponseWriter, r *http.Request) {
+	// Trace this method
+	ctx, span := trace.StartSpan(context.Background(), "ShopifyReceiver.Export")
```

Comment:
shouldn't this be r.Context()

Reply:
I don't think so. This is tracing ingestion internally, and we're not extending the caller's trace (if any). That said, the Zipkin receiver starts the span like this, but then adds a parent link if the context already has a parent span. 🤔

Reply:
the biggest problem that I've encountered with using the http.Request's context is that once the request is done it will cancel it.  Though we're not using deadlines etc.

Reply:
`trace.StartSpan` only uses the context to extract the parent span, if any, and returns a new context.WithValue containing the new span. Tracing itself doesn’t use the context or store it in the Span, so cancellation doesn’t really matter here. The key concern is whether the trace here should be a new one, with a parent link to the caller’s trace, or a continuation of the caller’s trace. The convention in other trace receivers seems to be a new one with a parent link, but I don’t know the rationale for that.

OK, so the original instrumentation was added & discussed in https://github.com/census-instrumentation/opencensus-service/issues/63. The justification for separate-but-linked traces is for bidirectional gRPC communication, but that doesn’t apply here. A TODO was added to the Zipkin receiver later to make this a direct descendant of the parent span rather than a link, so the right thing to do here is to pass `r.Context()` instead of `context.Background()`.